scripts/userdata/windows_bootstrap_userdata.txt

```diff
@@ -0,0 +1,31 @@
+<powershell>
+
+# USERDATA SCRIPT FOR AMAZON SOURCE WINDOWS SERVER AMIS
+# BOOTSTRAPS WINRM VIA SSL
+
+Set-ExecutionPolicy -ExecutionPolicy Unrestricted -Scope LocalMachine -Force -ErrorAction Ignore
+$ErrorActionPreference = "stop"
+
+# Remove any existing Windows Management listeners
+Remove-Item -Path WSMan:\Localhost\listener\listener* -Recurse
+
+# Create self-signed cert for encrypted WinRM on port 5986
+$Cert = New-SelfSignedCertificate -CertstoreLocation Cert:\LocalMachine\My -DnsName "packer-ami-builder"
+New-Item -Path WSMan:\LocalHost\Listener -Transport HTTPS -Address * -CertificateThumbPrint $Cert.Thumbprint -Force
+
+# Configure WinRM
+cmd.exe /c winrm quickconfig -q
+cmd.exe /c winrm set "winrm/config" '@{MaxTimeoutms="1800000"}'
+cmd.exe /c winrm set "winrm/config/winrs" '@{MaxMemoryPerShellMB="1024"}'
+cmd.exe /c winrm set "winrm/config/service" '@{AllowUnencrypted="false"}'
+cmd.exe /c winrm set "winrm/config/client" '@{AllowUnencrypted="false"}'
+cmd.exe /c winrm set "winrm/config/service/auth" '@{Basic="true"}'
+cmd.exe /c winrm set "winrm/config/client/auth" '@{Basic="true"}'
+cmd.exe /c winrm set "winrm/config/service/auth" '@{CredSSP="true"}'
+cmd.exe /c winrm set "winrm/config/listener?Address=*+Transport=HTTPS" "@{Port=`"5986`";Hostname=`"packer-ami-builder`";CertificateThumbprint=`"$($Cert.Thumbprint)`"}"
+cmd.exe /c netsh advfirewall firewall add rule name="WinRM-SSL (5986)" dir=in action=allow protocol=TCP localport=5986
+cmd.exe /c net stop winrm
+cmd.exe /c sc config winrm start= auto
+cmd.exe /c net start winrm
+
+</powershell>
```

Comment:
Is this needed for packer to use `winrm` to bootstrap windows machines? Can it instead pass scripts in via userdata?

Reply:
This script is the userdata script used to let packer use `winrm`, yes. We could in theory do something like concatenate all our scripts into userdata, but packer's delivery mechanism is configurable, and in the case of windows the preference is towards `winrm`.